### DISCUSSION TITLE
Dynamically generating JSON Schema with `const` properties

### DISCUSSION THREAD

**Discussion by aaronsgithub** [pydantic/pydantic#7966] [Question] [answered]
I want to create a base model, where subclasses have a generated field with a corresponding JSON schema property which is a `const`.

We can generate a `const` property in the JSON Schema by using a `Literal` type annotation on the Pydantic model field:
```
from pydantic import BaseModel, Field


class Component(BaseModel):
    type: str = Field(frozen=True)
    label: str

class TextArea(Component):
    type: typing.Literal["TextArea"] = Field(frozen=True)

> TextArea.model_json_schema()
{'properties': {'type': {'const': 'TextArea', 'title': 'Type'},
  'label': {'title': 'Label', 'type': 'string'}},
 'required': ['type', 'label'],
 'title': 'TextArea',
 'type': 'object'}
 ```
However, type checkers expect `Literal` to have a concrete value:
```
Type arguments for "Literal" must be None, a literal value (int, bool, str, or bytes), or an enum value
```
and so a dynamic expression cannot be used.

Is there an alternative to using `typing.Literal` which allows us to dynamically generate a pydantic model field with a  `const` propety in its JSON Schema?
```
import enum
import typing
from pydantic import BaseModel, Field

class Component(BaseModel):
    label: str

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)

        # I want to set type here dynamically for all subclasses so that it is a `const`
        # in the resulting json schema
        cls.type: ??? = ??? # = cls.__name__

class TextArea(Component):
    # some other fields
    pass

so that we get?

> TextArea.model_json_schema()
{'properties': {'type': {'const': 'TextArea', 'title': 'Type'},
  'label': {'title': 'Label', 'type': 'string'}},
 'required': ['type', 'label'],
 'title': 'TextArea',
 'type': 'object'}
```

**Comment by Viicos** [comment, marked as answer]
Using `enum.Enum` with a single value should also give you a `const` type, and enums can be build dynamically. If you still want to use `Literal`, you can also build it dynamically: `Literal[cls.__name__]  # type: ignore`

**Reply by aaronsgithub** [reply]
Dynamically created Enum works

**Comment by sydney-runkle** [comment]
`json_schema_extra` also might come in handy here: https://docs.pydantic.dev/dev/concepts/fields/#customizing-json-schema

**Reply by aaronsgithub** [reply]
Thanks, I've since dug into json_schema_extra and I realise I should be making better use of it.

It actually would be nicer to have type inside json_schema_extra so that it appears as an annotation instead of a property in the resulting JSON Schema.

However, I want to use type as a discrminator in discriminated union.
I may need to investigate if that is possible with a callable discriminator 🤔
